backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.cpp

```diff
@@ -0,0 +1,673 @@
+            for (auto annotation : key->annotations->annotations) {
+                if (annotation->name.name == "name") {
+                    if (annotation->body.size() > 0) annotationName = annotation->body[0]->text;
+                }
+            }
+            if (str.find(key->expression->toString()) != std::string::npos ||
+                (annotationName.size() > 0 && str.find(annotationName) != std::string::npos)) {
+                if (auto bit = key->expression->type->to<IR::Type_Bits>()) {
+                    type = bit;
+                } else if (auto varbit = key->expression->type->to<IR::Extracted_Varbits>()) {
+                    type = varbit;
+                } else if (key->expression->type->to<IR::Type_Unknown>()) {
+                    type = IR::Type_Bits::get(32);
+                } else {
+                    type = IR::Type_Bits::get(1);
+                }
+                result = IRUtils::getZombieConst(type, 0, str);
+                break;
+            }
+        }
+        if (!result) {
+            type = IR::Type_Bits::get(8);
+            result = IRUtils::getZombieConst(type, 0, str);
+        }
+
+    } else if (input.is(Token::Kind::Number)) {
+        if (!type) {
+            type = IR::Type_Bits::get(8);
+        }
+        result = IRUtils::getConstant(type, static_cast<big_int>(str));
+    }
+
+    return result;
+}
+
+/// Converts a vector of tokens into a single IR:Expression
+/// For example, at the input we have a vector of tokens:
+/// [key1(Text), ->(Implication), key2(Text), &&(Conjunction), key3(Text)] The result will be an
+/// IR::Expression equal to !IR::Expression || (IR::Expression && IR::Expression)
+const IR::Expression* getIR(std::vector<Token> tokens, IR::Vector<IR::KeyElement> keyElements) {
+    const IR::Expression* expr = nullptr;
+    std::vector<const IR::Expression*> exprVec;
+
+    for (int i = 0; i < (int)tokens.size(); i++) {
+        if (tokens[i].is_one_of(
+                Token::Kind::Minus, Token::Kind::Plus, Token::Kind::Equal, Token::Kind::NotEqual,
+                Token::Kind::GreaterThan, Token::Kind::GreaterEqual, Token::Kind::LessThan,
+                Token::Kind::LessEqual, Token::Kind::Slash, Token::Kind::Percent, Token::Kind::Shr,
+                Token::Kind::Shl, Token::Kind::Mul, Token::Kind::NotEqual)) {
+            const IR::Expression* leftL = nullptr;
+            const IR::Expression* rightL = nullptr;
+            if (i != 0) {
+                i--;
+                i--;
+            }
+            if (i <= 0 || tokens[i].is_one_of(Token::Kind::Conjunction, Token::Kind::Disjunction,
+                                              Token::Kind::LeftParen, Token::Kind::Implication)) {
+                i++;
+                leftL = makeConstant(tokens[i], keyElements);
+            } else {
+                leftL = exprVec[exprVec.size() - 1];
+                exprVec.pop_back();
+                i++;
+            }
+            i++;
```

Comment:
Please do not use index manipulation like this. It is confusing and error-prone. There is no need to have a single i++ or i-- in this entire file. I tried to refactor this but I do not quite understand the intention behind some of these statements. Furthermore, `i` must never be negative. That is considered a bug.
Also you are adding and removing expressions in this for loop from the vector you are iterating on at the same time. Please break up this code such that these two operations are cleanly separated.

Reply:
That is, to do something without changing the index?